File: Server/src/main/java/com/github/dr/rwserver/game/GameMaps.java
package com.github.dr.rwserver.game;

import com.github.dr.rwserver.data.global.Data;
import com.github.dr.rwserver.util.file.FileUtil;
import com.github.dr.rwserver.util.log.Log;
import com.github.dr.rwserver.util.zip.zip.ZipDecoder;

import static com.github.dr.rwserver.game.GameMaps.MapFileType.file;

/**
 * @author Dr
 */
public class GameMaps {

    public MapType mapType = MapType.defaultMap;
    public MapData mapData;
    public String mapName = "Crossing Large (10p)";
    public String mapPlayer = "[z;p10]";

    public static enum MapType {
        /**
         * ?
         */
        defaultMap, customMap, savedGames
    }

    public static enum MapFileType {
        /**
         * ?
         */
        file, zip, web
    }

    public static class MapData {
        public final MapType mapType;
        public final MapFileType mapFileType;
        public final String mapFileName;
        public final String zipFileName;

        public int mapSize = 0;
        public byte[] bytesMap = null;

        public MapData(final MapType mapType,final MapFileType mapFileType,final String mapFileName) {
            this.mapType = mapType;
            this.mapFileType = mapFileType;
            this.mapFileName = mapFileName;
            this.zipFileName = null;
        }

        public MapData(final MapType mapType,final MapFileType mapFileType,final String mapFileName,final String zipFileName) {
            this.mapType = mapType;
            this.mapFileType = mapFileType;
            this.mapFileName = mapFileName;
            this.zipFileName = zipFileName;
        }

        public MapData(final MapType mapType,final byte[] bytesMap) {
            this.mapType = mapType;
            this.mapFileType = file;
            this.mapFileName = null;
            this.zipFileName = null;
            this.mapSize = bytesMap.length;
            this.bytesMap = bytesMap;
        }

        public MapData(final MapType mapType,final int mapSize,final byte[] bytesMap) {
            this.mapType = mapType;
            this.mapFileType = file;
            this.mapFileName = null;
            this.zipFileName = null;
            this.mapSize = mapSize;
            this.bytesMap = bytesMap;
        }

        public String getType() {
            return mapType.name().equals("savedGames") ? ".save" : ".tmx";
        }

        public void readMap() {
            FileUtil fileUtil = FileUtil.File(Data.Plugin_Maps_Path);
            switch (mapFileType) {
                case file:
                    try {
                        this.bytesMap = fileUtil.toPath(mapFileName+getType()).readFileByte();
                        this.mapSize = bytesMap.length;
                    } catch (Exception e) {
                        Log.error("Read Map Bytes Error",e);
                    }
                    break;
                case zip:
                    try {
                        this.bytesMap = new ZipDecoder(fileUtil.toPath(zipFileName).getFile()).GetTheFileBytesOfTheSpecifiedSuffixInTheZip(this);
                        this.mapSize = bytesMap.length;
                    } catch (Exception e) {
                        Log.error("Read Map Bytes Error",e);
                    }
                    break;
                case web:
                    break;
                default:
                    break;
            }
        }
    }
}

File: Server/src/main/java/com/github/dr/rwserver/game/GameMaps.kt
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package com.github.dr.rwserver.game

import com.github.dr.rwserver.data.global.Data
import com.github.dr.rwserver.util.file.FileUtil.Companion.getFolder
import com.github.dr.rwserver.util.log.Log.error
import com.github.dr.rwserver.util.zip.zip.ZipDecoder

/**
 * 游戏的地图数据
 * @author Dr
 */
class GameMaps {
    /** 地图类型  */
    @JvmField
    var mapType = MapType.defaultMap

    /** 地图数据  */
    @JvmField
    var mapData: MapData? = null

    /** 地图名  */
    @JvmField
    var mapName = "Crossing Large (10p)"

    /** 地图人数  */
    @JvmField
    var mapPlayer = "[z;p10]"

    /** 地图的类型  */
    enum class MapType {
        /**
         * defaultMap : 官方地图
         * customMap  : 自定义的地图
         * savedGames : 游戏里保存的进度
         */
        defaultMap, customMap, savedGames
    }

    enum class MapFileType {
        /**
         * file : 以文件的形式提供
         * zip  : 以ZIP提供(文件在ZIP内)
         * web  : 以WebFile的形式提供(需要下载)
         */
        file, zip, web
    }

    class MapData {
        @JvmField
        val mapType: MapType
        @JvmField
        val mapFileType: MapFileType
        @JvmField
        val mapFileName: String
        @JvmField
        val zipFileName: String?
        @JvmField
        var mapSize = 0
        @JvmField
        var bytesMap: ByteArray? = null

        constructor(mapType: MapType, mapFileType: MapFileType, mapFileName: String) {
            this.mapType = mapType
            this.mapFileType = mapFileType
            this.mapFileName = mapFileName
            zipFileName = null
        }

        constructor(mapType: MapType, mapFileType: MapFileType, mapFileName: String, zipFileName: String) {
            this.mapType = mapType
            this.mapFileType = mapFileType
            this.mapFileName = mapFileName
            this.zipFileName = zipFileName
        }

        /**
         * 根据地图类型获取文件后缀
         * @return 后缀
         */
        val type: String
            get() = if ("savedGames" == mapType.name) ".save" else ".tmx"

        /**
         * 懒加载
         * 读取地图到内部bytes
         */
        fun readMap() {
            val fileUtil = getFolder(Data.Plugin_Maps_Path)
            when (mapFileType) {
                MapFileType.file -> try {
                    bytesMap = fileUtil.toFile(mapFileName + type).readFileByte()
                    mapSize = bytesMap!!.size
                } catch (e: Exception) {
                    error("Read Map Bytes Error", e)
                }
                MapFileType.zip -> try {
                    bytesMap =
                        ZipDecoder(fileUtil.toFile(zipFileName!!).file).getTheFileBytesOfTheSpecifiedSuffixInTheZip(this)
                    mapSize = bytesMap!!.size
                } catch (e: Exception) {
                    error("Read Map Bytes Error", e)
                }
                MapFileType.web -> {
                }
            }
        }

        /**
         * 清理服务器使用的地图数据
         */
        fun clean() {
            mapSize = 0
            bytesMap = null
        }
    }
}